### PR TITLE
Assignment5

### PR DESCRIPTION
**Pull request by Sithale**
Did not do extra credit.

### CODE REVIEW

**Comment by cmartinezaraujo** [comment]
I will review.

File: GenericsHomework/GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        public Node<N> Next
+        {
+            get => _Next!;
+            set { value._Next = this; _Next = value; }
+
+        }
+
+        public Node(N node)
+        {
+            this.Data = node;
```
Comment (cmartinezaraujo) [inline GenericsHomework/GenericsHomework/Node.cs:20]: Remember to take out the "this" keyword in places where it Is not absolutely needed.

**Review by cmartinezaraujo — COMMENTED** [review]
- Create a class library project called "GenericsHomework.". ✔

Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.

Define the Node class
- Include a constuctor that takes a value. (No validation is necessary on the value). ✔
- Add a ToString() override that writes out the value's ToString() result. ✔
- Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
- The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
- The Next property setter should be private. ✔
- In addition to non-null validation, the body of the Next property should insert the next Node (the value) into the list. ✔
- Add an Insert method that takes a value and inserts a new Node instance after the current node (by invoking the Next property). ✔

Add a clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:

- Whether it is sufficient to only set Next to itself ✔
- Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
- Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage colleciton because all the items point to each other and therefore may never be garbage collected. ✔
- You should not rely on any BCL generic classes for your implementation. ✔

Extra Credit:
1. Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌
2. Implement Systm.Collections.Generic.ICollection<T> on the Node class ❌

Fundamentals:

- Ensure you enable nullable reference types, net5 targetted, C# 9.0, warnings as errors, and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEquals<T>() (the generic version)
- All of the above should be unit tested ✔

**Comment by mmwoodfo** [comment]
When doing your code reviews it is more beneficial for the person you are reviewing to give them inline comments (similar to what you see me doing on your PR) rather than a single comment listing all of the issues. This isn't something you are going to be dinged for on this assignment, but you may lose points on future assignments for unclear or difficult to read reviews.'

It's fine to keep the large block comment with the full checklist of items done well and not done well, but in the future try to give inline comments for any 'X's that you leave on the review rather than adding the reasoning to the block of text.

**Comment by Sithale** [comment]
Ok, thank you! I will be sure to add some more input on future reviews. I also noticed that people would start with "I will review this", and I had not done that. So, I will be doing that as well now too.
________________________________

File: GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            double num = 19.99;
+            string str = "Sloppy";
+
+            // Act
+            Node<double> node1 = new Node<double>(num);
+            Node<string> node2 = new Node<string>(str);
+
+            // Assert
+            Assert.AreEqual<string>(num.ToString(), node1.ToString());
+            Assert.AreEqual<string>(str.ToString(), node2.ToString());
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework.Tests/NodeTests.cs:39]:
```suggestion
            Assert.AreEqual(num.ToString(), node1.ToString());
            Assert.AreEqual(str.ToString(), node2.ToString());
```

File: GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        {
+
+            // Act
+            Node<string> node1 = new Node<string>("Howdy");
+            node1.Insert("Love");
+
+            // Assert
+            Assert.AreEqual<string>("Howdy", node1.ToString());
+            Assert.AreEqual<string>("Love", node1.Next.ToString());
+            Assert.AreEqual<string>("Howdy", node1.Next.Next.ToString());
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework.Tests/NodeTests.cs:92]: Try to avoid using your `ToString` method when unit testing unrelated code. Your unit tests should be focused on a single _unit_ of your code, in this case, just the `Insert` method.

File: GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+        }
+
+        [TestMethod]
+        public void Node_ClearFunction_AllNodesExceptFirstOneAreCleared()
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework.Tests/NodeTests.cs:97]: You may consider writing a unit test to test that the garbage collector actually cleans up your detached nodes.
